src/design_docs/async_stack_traces.md

```diff
@@ -0,0 +1,705 @@
+   2: core::panicking::panic
+            at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\core\src\panicking.rs:50
+   3: common::baz::generator$0
+            at C:\Users\ericholk\repo\backtrace-examples\async-common\src\lib.rs:10
+   4: core::future::from_generator::impl$1::poll<common::baz::generator$0>
+            at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\core\src\future\mod.rs:80
+   5: common::bar::generator$0
+            at C:\Users\ericholk\repo\backtrace-examples\async-common\src\lib.rs:6
+   6: core::future::from_generator::impl$1::poll<common::bar::generator$0>
+            at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\core\src\future\mod.rs:80
+   7: common::foo::generator$0
+            at C:\Users\ericholk\repo\backtrace-examples\async-common\src\lib.rs:2
+   8: core::future::from_generator::impl$1::poll<common::foo::generator$0>
+            at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\core\src\future\mod.rs:80
+   9: async_tokio::main::generator$0
+            at .\src\main.rs:4
+```
+
+To achieve parity with the synchronous backtrace, there are two more improvements needed.
+
+The first is that in between each frame in user code, there is a frame with a call to `core::future::from_generator`.
+
+The second is that rather than seeing a call to `common::foo` or similar, we see `common::foo::generator$0`.
+
+These two are relatively minor issues.
+Fixing them may not be desirable.
+
+## Proposed Solutions
+
+There are several improvements we could make that would improve the state of backtraces.
+
+### Allow runtimes to trim startup code
+
+This solution probably has the highest impact to effort ratio.
+The core of the idea is to move the call to `__rust_begin_short_backtrace` when running under an async executor.
+Most likely the way this would happen is to have an alternate startup path that programs can opt in to (how to actually make this work is left as an exercise for the reader).
+The alternate path would not call `__rust_begin_short_backtrace` in [`rt.rs`], but would instead expect the program to make sure to call it at the appropriate time.
+Then, async runtimes that provide a macro such as `#[tokio::main]` or `#[async_std::main]` would arrange to start up through this alternate path and call `__rust_begin_short_backtrace` shortly before invoking user code.
+
+[`rt.rs`]: https://github.com/rust-lang/rust/blob/master/library/std/src/rt.rs
+
+Note that this approach does not help in cases where users create and launch the runtime manually rather than using a library-provided macro.
+It does help in the most common cases, however, and when users take a more manual approach they would also be able to control where the short stack trace starts.
+
+### Allow trimming of internal frames
+
+There are several ways to do this, with varying levels of implementation effort and cost.

```

Comment:
These are pretty implementation-focussed. My preferred solution is that any function can have an attribute, say `#[backtrace_transparent]`, and then it is elided from the default backtrace. Of course, how this is implemented is a hard question; I think either of your first two would work, but neither feel great.

Reply:
I like that idea too. I'll add a note describing that idea.

It occurs to me that we could probably experiment with some of these ideas with an external crate before we commit to doing this in `std` or `rustc`.